src/mol-util/array.ts

```diff
@@ -156,3 +158,73 @@ export function arrayMapUpsert<T>(xs: [string, T][], key: string, value: T) {
     }
     xs.push([key, value]);
 }
+
+/** Return an array containing integers from [start, end) if `end` is given,
+ * or from [0, start) if `end` is omitted. */
+export function range(start: number, end?: number): number[] {
+    if (end === undefined) {
+        end = start;
+        start = 0;
+    }
+    const length = Math.max(end - start, 0);
+    const result = Array(length);
+    for (let i = 0; i < length; i++) {
+        result[i] = start + i;
+    }
+    return result;
+}
+
+/** Copy all elements from `src` to the end of `dst`.
+ * Equivalent to `dst.push(...src)`, but avoids storing element on call stack. Faster that `extend` from Underscore.js.
+ * `extend(a, a)` will double the array.
+ */
+export function arrayExtend<T>(dst: T[], src: ArrayLike<T>): void {
+    const offset = dst.length;
+    const nCopy = src.length;
+    dst.length += nCopy;
+    for (let i = 0; i < nCopy; i++) {
+        dst[offset + i] = src[i];
+    }
+}
+
+/** Check whether `array` is sorted, sort if not. */
```

Comment:
Surprisingly, when the array is already sorted, this can be ~3x faster.
I use it when I expect the array will usually be sorted, but have to double-check (e.g. auth_seq_id).

Reply:
cool, would you mind adding a little test to perf-tests/

Reply:
@dsehnal already merged this PR, so I'll address it a future one (https://github.com/molstar/molstar/pull/991)

Reply:
Addressed in this commit (https://github.com/molstar/molstar/pull/991/commits/1d200c9a8b969965b934e0b68888cd371f540996)

This is what the perf test gives:
```
sort (1000000, pre-sorted): 37.778ms
sort (1000000, pre-sorted): 38.496ms
sort (1000000, pre-sorted): 39.071ms
sortIfNeeded (1000000, pre-sorted): 2.355ms
sortIfNeeded (1000000, pre-sorted): 2.99ms
sortIfNeeded (1000000, pre-sorted): 0.992ms
sort (1000000, not sorted): 459.946ms
sort (1000000, not sorted): 422.959ms
sort (1000000, not sorted): 418.736ms
sortIfNeeded (1000000, not sorted): 437.739ms
sortIfNeeded (1000000, not sorted): 382.919ms
sortIfNeeded (1000000, not sorted): 376.24ms
```